src/mongoose_rabbit_worker.erl

```diff
@@ -168,9 +187,35 @@ make_chat_msg(From, To, UserMsg) ->
 is_user_online(online) -> true;
 is_user_online(offline) -> false.
 
--spec close_rabbit_connection(Connection :: pid(), Channel :: pid()) ->
+-spec publish_message_and_wait_for_confirm(Channel :: pid(),
+                                           Exchange :: binary(),
+                                           RoutingKey :: binary(),
+                                           Message :: binary()) -> term().
+publish_message_and_wait_for_confirm(Channel, Exchange, RoutingKey, Message) ->
+    amqp_channel:call(Channel, #'basic.publish'{exchange = Exchange,
+                                                routing_key = RoutingKey},
+                      #amqp_msg{payload = Message}),
+    amqp_channel:wait_for_confirms(Channel).
+
+-spec establish_rabbit_connection(Opts :: #amqp_params_network{},
+                                  Host :: jid:server()) -> ok.
+establish_rabbit_connection(Opts, Host) ->
+    {ok, Connection} = amqp_connection:start(Opts),
+    {ok, Channel} = amqp_connection:open_channel(Connection),
+    mongoose_metrics:update(Host, ?RABBIT_CONNECTIONS_METRIC, 1),
+    ?DEBUG("event=rabbit_connection_established", []),
+    {Connection, Channel}.
+
+-spec enable_confirms(Channel :: pid()) -> ok.
+enable_confirms(Channel) ->
+    #'confirm.select_ok'{} = amqp_channel:call(Channel, #'confirm.select'{}),
+    ok.
+
+-spec close_rabbit_connection(Connection :: pid(), Channel :: pid(),
+                              Host :: jid:server()) ->
     ok | term().
-close_rabbit_connection(Connection, Channel) ->
+close_rabbit_connection(Connection, Channel, Host) ->
+    mongoose_metrics:update(Host, ?RABBIT_CONNECTIONS_METRIC, -1),
     try amqp_channel:close(Channel)
     catch
         _Error:_Reason -> already_closed
```

Comment:
are we sure that "already closed" is really error here?

Reply:
Could you please explain what you meant here? We just try to close a channel and don't care about the result. We have to catch a possible exception cause if such occurs we won't be able to close a rabbit connection cause a worker process will die on this call.

Reply:
hmm but why there might be exception?
If so, can we log it with `?ERROR( ...` ?
my point is that with this implementation, the error will be completely ignored

Reply:
There might be an exception because e.g. a Rabbit server closed a connection and we try to close the connection that is already closed, so there is no need to log this as error.

Reply:
IS it possible to distinguish this kind of exception ? _Error:_Reason matches on everything

Reply:
As far as I know this call causes an exception only if the channel we're trying to close is already closed or if some timeout is exceeded (it's `gen_server:call` underneath).